### PR TITLE
T4 - GCD & LCM Generator

### PR DESCRIPTION
**Pull request by vikrampg1710-tmm**

### CODE REVIEW

File: Assignments/Program.cs [hunk 1]
```diff
@@ -1 +1,77 @@
...
+      if (a < b) (a, b) = (b, a);
+      return (a == 0) ? b : ((b == 0) ? a : EuclidGCDof (b, a % b));
+   }
+
+   /// <summary>Returns the LCM of input two numbers using Euclidean Algorithm</summary>
+   public static int EuclidLCMof (int a, int b) 
+      => Math.Abs (a* b) / EuclidGCDof (b, a % b);
+
+   public static void WriteInYellow (string output) {
+      Console.WriteLine (output, Console.ForegroundColor = Yellow);
```
Comment (sdinesh-90) [inline Assignments/Program.cs]: Can you explain what you are doing here? Which overload of WriteLine you are using?

File: Assignments/Program.cs [hunk 2]
```diff
@@ -1 +1,77 @@
...
+
+   /// <summary>Returns the GCD of input two numbers using Euclidean Algorithm</summary>
+   public static int EuclidGCDof (int a, int b) {
+      if (a < b) (a, b) = (b, a);
+      return (a == 0) ? b : ((b == 0) ? a : EuclidGCDof (b, a % b));
+   }
+
+   /// <summary>Returns the LCM of input two numbers using Euclidean Algorithm</summary>
+   public static int EuclidLCMof (int a, int b) 
+      => Math.Abs (a* b) / EuclidGCDof (b, a % b);
```
Comment (sdinesh-90) [inline Assignments/Program.cs]:
```suggestion
      => Math.Abs (a * b) / EuclidGCDof (b, a % b);

```

File: Assignments/Program.cs [hunk 3]
```diff
@@ -1 +1,77 @@
...
+         if (inputs.All (a => value % a == 0))
+            return value;
+      }
+      return 1;
+   }
+
+   /// <summary>Returns the GCD of input two numbers using Euclidean Algorithm</summary>
+   public static int EuclidGCDof (int a, int b) {
+      if (a < b) (a, b) = (b, a);
+      return (a == 0) ? b : ((b == 0) ? a : EuclidGCDof (b, a % b));
```
Comment (sdinesh-90) [inline Assignments/Program.cs]:
```suggestion
      return a == 0 ? b : (b == 0 ? a : EuclidGCDof (b, a % b));

```

File: Assignments/Program.cs [hunk 4]
```diff
@@ -1 +1,77 @@
...
+         Console.ResetColor ();
+      }
+   }
+
+   /// <summary>Returns the Greatest Common Factor of all numbers from a input list</summary>
+   public static int GCDof (List<int> inputs) {
+      if (inputs.Count == 2) return EuclidGCDof (inputs[0], inputs[1]);
+      for (int value = inputs.Min (); value > 0; value--) {
+         if (inputs.All (a => a % value == 0))
+            return value;
```
Comment (sdinesh-90) [inline Assignments/Program.cs]: If user enter 222222, 222223, and 222224, for loop will run for 222222 times. This is not necessary, you could get GCD of two input and then try GCD of that result and third input.

**Review by sdinesh-90 — CHANGES_REQUESTED** [review]

File: Assignments/Program.cs (hunk 1 above)
Comment (vikrampg1710-tmm) [reply Assignments/Program.cs]: Is this the overload of WriteLine?
public static void WriteLine([StringSyntax(StringSyntaxAttribute.CompositeFormat)] string format, object? arg0)
        {
            Out.WriteLine(format, arg0);
        }

**Comment by vikrampg1710-tmm** [comment]
@sdinesh-90 All your above suggestions have been implemented in the next commit 2d5ecd7.  Please check.

File: Assignments/Program.cs (hunk 1 above)
Comment (sdinesh-90) [reply Assignments/Program.cs]: This is of format Console.WriteLine ("Test {0}", "1"). But you are not using that way.

File: Assignments/Program.cs [hunk 5]
```diff
@@ -1 +1,86 @@
...
+
+namespace Spark;
+
+public class T4 {
+   public static void Main () {
+      Console.WriteLine ("\x1B[4m" + "GCD & LCM Generator:-" + "\x1B[0m");
+      List<List<int>> testCases = new () { new () { 5, 10, 15, 21 }, new () { 6, 18, 36, 54, 96 }, 
+         new () { 6, 8, 12, 10, 30, 72, 134 }, new () { 2, 3, 5, 7, 11, 13 }, new () { 17, 34, 51, 68, 36 } };
+      for (int i = 0; i < testCases.Count; i++) {
+         Console.Write ($"\nTest case-{i + 1}: ");
```
Comment (sdinesh-90) [inline Assignments/Program.cs]: Don't use \n as line ending. See [here](https://www.cs.toronto.edu/~krueger/csc209h/tut/line-endings.html#:~:text=Text%20files%20created%20on%20DOS,(%22%5Cn%22).).

Comment (sdinesh-90) [reply Assignments/Program.cs]: Change in other places also.

File: Assignments/Program.cs [hunk 6]
```diff
@@ -1 +1,86 @@
...
+      Console.Write ("\nGCD = "); WriteInYellow ($"{GCDof (input)}");
+      Console.Write ("LCM = "); WriteInYellow ($"{LCMof (input)}");
+   }
+
+   /// <summary>Returns the Greatest Common Factor of all the numbers in the input list</summary>
+   public static int GCDof (List<int> inputs) {
+      int result = EuclidGCDof (inputs[0], inputs[1]);
+      for (int i = 2; i < inputs.Count; i++) {
+         var temp = EuclidGCDof (result, inputs[i]);
+         result = temp;
```
Comment (sdinesh-90) [inline Assignments/Program.cs]: temp is not required.

File: Assignments/Program.cs [hunk 7]
```diff
@@ -1 +1,86 @@
...
+      }
+      return result;
+   }
+   
+   /// <summary>Returns the Least Common Multiple of all numbers in the input list</summary>
+   public static int LCMof (List<int> inputs) {
+      int result = EuclidLCMof (inputs[0], inputs[1]);
+      for (int i = 2; i < inputs.Count; i++) {
+         int temp = EuclidLCMof (result, inputs[i]);
+         result = temp;
```
Comment (sdinesh-90) [inline Assignments/Program.cs]: temp variable not required.

**Review by sdinesh-90 — CHANGES_REQUESTED** [review]

**Comment by vikrampg1710-tmm** [comment]
@sdinesh-90 All the above-mentioned suggestions have been implemented in the following commit b7dcc40. Please check.

**Review by sdinesh-90 — APPROVED** [review]

**Pull request closed without merging**